cmd/virt-handler/virt-handler.go

```diff
@@ -236,6 +331,62 @@ func (app *virtHandlerApp) AddFlags() {
 		"Number of devices to register with Kubernetes device plugin framework")
 }
 
+func (app *virtHandlerApp) setupTLS() error {
+
+	clientCert, err := tls.X509KeyPair(app.clientCertBytes, app.clientKeyBytes)
+	if err != nil {
+		return err
+	}
+
+	caCert, err := cert.ParseCertsPEM(app.signingCertBytes)
+	if err != nil {
+		return err
+	}
+
+	certPool := x509.NewCertPool()
+
+	for _, crt := range caCert {
+		certPool.AddCert(crt)
+	}
+
+	app.migrationTLSConfig = &tls.Config{
+		MinVersion: tls.VersionTLS12,
+		GetClientCertificate: func(info *tls.CertificateRequestInfo) (certificate *tls.Certificate, e error) {
+			return &clientCert, nil
+		},
+		GetCertificate: func(info *tls.ClientHelloInfo) (i *tls.Certificate, e error) {
+			return &clientCert, nil
+		},
+		// Neither the client nor the server should validate anything itself, `VerifyPeerCertificate` is still executed
+		InsecureSkipVerify: true,
+		// XXX: We need to verify the cert ourselves because we don't have DNS or IP on the certs at the moment
+		VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+
+			// impossible with RequireAnyClientCert
+			if len(rawCerts) == 0 {
+				return fmt.Errorf("no client certificate provided.")
+			}
+
+			c, err := x509.ParseCertificate(rawCerts[0])
+			if err != nil {
+				return fmt.Errorf("failed to parse peer certificate: %v", err)
+			}
+			_, err = c.Verify(x509.VerifyOptions{
+				Roots:     certPool,
+				KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+			})
+
+			if err != nil {
+				return fmt.Errorf("could not verify peer certificate: %v", err)
+			}
+			return nil
+		},
+		ClientAuth: tls.RequireAnyClientCert,
```

Comment:
RequireAndVerifyClientCert instead?

Reply:
Changed. Luckily it makes no difference. In both cases the verify function is called.